src/controls/controls.actions.ts

```diff
@@ -268,138 +270,116 @@ import { renderCircleControl, renderSquareControl } from './controls.render';
+      const { skewX, scaleX } = target,
+        dimBefore = target._getTransformedDimensions({ skewY: skewingBefore }),
+        dimAfter = target._getTransformedDimensions(),
+        compensationFactor = skewX !== 0 ? dimBefore.x / dimAfter.x : 1;
+      compensationFactor !== 1 &&
+        target.set('scaleX', compensationFactor * scaleX);
+    }
+
+    return changed;
   }
 
   /**
-   * Action handler for skewing on the X axis
-   * @private
+   * Wrapped Action handler for skewing on a given axis, takes care of the
+   * skew direction and determine the correct transform origin for the anchor point
+   * @param {Event} eventData javascript event that is doing the transform
+   * @param {Object} transform javascript object containing a series of information around the current transform
+   * @param {number} x current mouse x position, canvas normalized
+   * @param {number} y current mouse y position, canvas normalized
+   * @return {Boolean} true if some change happened
    */
-  function skewObjectX(eventData, transform, x, y) {
-    var target = transform.target,
-      // find how big the object would be, if there was no skewX. takes in account scaling
-      dimNoSkew = target._getTransformedDimensions({
-        skewX: 0,
-        skewY: target.skewY,
-      }),
-      localPoint = getLocalPoint(
-        transform,
-        transform.originX,
-        transform.originY,
-        x,
-        y
-      ),
-      // the mouse is in the center of the object, and we want it to stay there.
-      // so the object will grow twice as much as the mouse.
-      // this makes the skew growth to localPoint * 2 - dimNoSkew.
-      totalSkewSize = Math.abs(localPoint.x * 2) - dimNoSkew.x,
-      currentSkew = target.skewX,
-      newSkew;
-    if (totalSkewSize < 2) {
-      // let's make it easy to go back to position 0.
-      newSkew = 0;
-    } else {
-      newSkew = radiansToDegrees(
-        Math.atan2(totalSkewSize / target.scaleX, dimNoSkew.y / target.scaleY)
-      );
-      // now we have to find the sign of the skew.
-      // it mostly depend on the origin of transformation.
-      if (transform.originX === LEFT && transform.originY === BOTTOM) {
-        newSkew = -newSkew;
-      }
-      if (transform.originX === RIGHT && transform.originY === TOP) {
-        newSkew = -newSkew;
-      }
-      if (targetHasOneFlip(target)) {
-        newSkew = -newSkew;
-      }
-    }
-    var hasSkewed = currentSkew !== newSkew;
-    if (hasSkewed) {
-      var dimBeforeSkewing = target._getTransformedDimensions().y;
-      target.set('skewX', newSkew);
-      compensateScaleForSkew(target, 'skewY', 'scaleY', 'y', dimBeforeSkewing);
+  function skewHandler(axis: TAxis, eventData, transform, x, y) {
+    const { target } = transform,
+      {
+        counterAxis,
+        origin: originKey,
+        lockSkewing: lockSkewingKey,
+        skew: skewKey,
+      } = AXIS_KEYS[axis];
+    if (target[lockSkewingKey]) {
+      return false;
```

Comment:
we should probably move all of those in its own wrapper if possible, something like `lockedBy('prop', handler)` just an idea for a task. We have scale, skew, rotate that all have the same concept

Reply:
In general controls need es6 + TS work
That can be added there